File: node/src/providers/ollama-adapter.ts
import type {
  Message,
  ProviderUsage,
  StreamEvent,
  TextDelta,
  ToolCallEvent,
  UsageEvent,
} from "../types.js"
import type {
  CanonicalAdapterInput,
  CanonicalMessage,
} from "./content-normalization.js"
import { projectToolOutputToText } from "./content-normalization.js"
import { normalizeToolCall } from "./base.js"
import {
  type AdapterDecodeInput,
  type AdapterOutput,
  type AdapterStreamInput,
  type CanonicalStopReason,
  type ProtocolAdapter,
  ProtocolResponseError,
  OLLAMA_PROTOCOL_CAPABILITIES,
} from "./protocol-adapter.js"

export interface OllamaMessage {
  role: string
  content: string
  images?: string[]
}

export interface OllamaRequest {
  model: string
  messages: OllamaMessage[]
  tools?: Array<{
    type: "function"
    function: { name: string; description: string; parameters: unknown }
  }>
  [key: string]: unknown
}

export interface OllamaChunk {
  message?: {
    content?: string
    tool_calls?: Array<{
      function: { name: string; arguments: unknown }
    }>
  }
  done?: boolean
  done_reason?: string
  prompt_eval_count?: number
  eval_count?: number
}

export interface OllamaStreamState {
  readonly input: CanonicalAdapterInput
  readonly pendingToolCalls: Map<string, {
    id: string
    name: string
    arguments: Record<string, unknown>
  }>
  finalChunk?: OllamaChunk
}

function messageContent(message: CanonicalMessage): OllamaMessage {
  const text: string[] = []
  const images: string[] = []
  for (const item of message.blocks) {
    if (item.type === "tool_result") {
      text.push(projectToolOutputToText(item.blocks))
    } else if (item.type === "text") {
      text.push(item.text)
    } else if (item.type === "image" && item.source.kind === "base64") {
      images.push(item.source.data)
      text.push("[image]")
    } else {
      throw new ProtocolResponseError("ollama-chat", `cannot serialize ${item.type}`)
    }
  }
  return {
    role: message.role,
    content: text.join("\n"),
    ...(images.length ? { images } : {}),
  }
}

function requestExtensions(
  extensions: Readonly<Record<string, unknown>>,
): Record<string, unknown> {
  const blocked = new Set([
    "model",
    "messages",
    "tools",
    "stream",
    "__deepstrikeThinkingEnabled",
    "degradeMissingReasoningReplay",
  ])
  return Object.fromEntries(
    Object.entries(extensions).filter(([key]) => !blocked.has(key)),
  )
}

function validCount(raw: Record<string, unknown>, field: string): number | undefined {
  const value = raw[field]
  if (value === undefined) return undefined
  if (typeof value !== "number" || !Number.isFinite(value) || value < 0) {
    throw new ProtocolResponseError(
      "ollama-chat",
      `${field} must be a non-negative finite number`,
    )
  }
  return value
}

export class OllamaNdjsonDecoder {
  private buffer = ""

  push(text: string): OllamaChunk[] {
    this.buffer += text
    const lines = this.buffer.split("\n")
    this.buffer = lines.pop() ?? ""
    return this.parse(lines)
  }

  finish(text = ""): OllamaChunk[] {
    this.buffer += text
    const tail = this.buffer
    this.buffer = ""
    return this.parse(tail ? [tail] : [])
  }

  private parse(lines: readonly string[]): OllamaChunk[] {
    const chunks: OllamaChunk[] = []
    for (const line of lines) {
      if (!line.trim()) continue
      try {
        const value = JSON.parse(line) as unknown
        if (value && typeof value === "object" && !Array.isArray(value)) {
          chunks.push(value as OllamaChunk)
        }
      } catch {
        // Preserve the established Ollama behavior: malformed complete lines are skipped.
      }
    }
    return chunks
  }
}

export class OllamaAdapter implements ProtocolAdapter<
  OllamaRequest,
  OllamaChunk,
  OllamaChunk,
  OllamaStreamState,
  OllamaChunk | undefined
> {
  readonly protocol = "ollama-chat" as const
  readonly protocolCapabilities = OLLAMA_PROTOCOL_CAPABILITIES

  buildRequest(input: CanonicalAdapterInput): OllamaRequest {
    const messages: OllamaMessage[] = []
    if (input.context.systemText) {
      messages.push({ role: "system", content: input.context.systemText })
    }
    const turns = input.context.stateTurn
      ? [...input.context.turns, input.context.stateTurn]
      : input.context.turns
    messages.push(...turns.map(messageContent))
    return {
      ...requestExtensions(input.extensions),
      model: input.resolved.identity.modelId,
      messages,
      ...(input.tools.length ? {
        tools: input.tools.map(tool => ({
          type: "function" as const,
          function: {
            name: tool.name,
            description: tool.description,
            parameters: JSON.parse(tool.parameters) as unknown,
          },
        })),
      } : {}),
    }
  }

  decodeComplete(raw: OllamaChunk, _input: AdapterDecodeInput): { message: Message } {
    return {
      message: {
        role: "assistant",
        content: raw.message?.content ?? "",
      },
    }
  }

  createStreamState(input: AdapterStreamInput): OllamaStreamState {
    return { input: input.input, pendingToolCalls: new Map() }
  }

  pushStreamChunk(chunk: OllamaChunk, state: OllamaStreamState): AdapterOutput {
    const events: StreamEvent[] = []
    if (chunk.message?.content) {
      events.push({ type: "text_delta", delta: chunk.message.content } as TextDelta)
    }
    for (const call of chunk.message?.tool_calls ?? []) {
      const normalized = normalizeToolCall(
        "",
        call.function.name,
        call.function.arguments,
      )
      if (!normalized) continue
      const key = `${normalized.name}:${normalized.arguments}`
      if (!state.pendingToolCalls.has(key)) {
        state.pendingToolCalls.set(key, {
          id: `call_${state.pendingToolCalls.size + 1}`,
          name: normalized.name,
          arguments: JSON.parse(normalized.arguments) as Record<string, unknown>,
        })
      }
    }
    if (chunk.done) state.finalChunk = chunk
    return { events }
  }

  finishStream(
    state: OllamaStreamState,
    final: OllamaChunk | undefined,
  ): AdapterOutput {
    const terminal = final ?? state.finalChunk
    const events: StreamEvent[] = Array.from(
      state.pendingToolCalls.values(),
      call => ({ type: "tool_call", ...call }) as ToolCallEvent,
    )
    const usage = this.normalizeUsage(terminal)
    if (usage) {
      const rawStopReason = terminal?.done_reason
      const stopReason = this.normalizeStopReason(rawStopReason)
      events.push({
        type: "usage",
        totalTokens: usage.inputTokens + usage.outputTokens,
        inputTokens: usage.inputTokens,
        outputTokens: usage.outputTokens,
        providerUsage: usage,
        ...(stopReason ? { stopReason } : {}),
        ...(rawStopReason ? { rawStopReason } : {}),
      } as UsageEvent)
    }
    return { events }
  }

  normalizeUsage(raw: unknown): ProviderUsage | undefined {
    if (raw === undefined || raw === null) return undefined
    if (typeof raw !== "object" || Array.isArray(raw)) {
      throw new ProtocolResponseError("ollama-chat", "usage source must be an object")
    }
    const record = raw as Record<string, unknown>
    const inputTokens = validCount(record, "prompt_eval_count")
    const outputTokens = validCount(record, "eval_count")
    if (inputTokens === undefined && outputTokens === undefined) return undefined
    return {
      inputTokens: inputTokens ?? 0,
      outputTokens: outputTokens ?? 0,
    }
  }

  normalizeStopReason(raw: string | undefined): CanonicalStopReason | undefined {
    if (raw === undefined) return undefined
    switch (raw) {
      case "stop": return "end_turn"
      case "length": return "max_tokens"
      default: return "other"
    }
  }

  createNdjsonDecoder(): OllamaNdjsonDecoder {
    return new OllamaNdjsonDecoder()
  }
}

File: python/deepstrike/providers/ollama_adapter.py
"""Ollama request conversion plus NDJSON stream lifecycle."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from deepstrike._kernel import Message, ToolSchema
from deepstrike.providers.base import UnsupportedModalityError, normalize_tool_call
from deepstrike.providers.protocol_adapter import AdapterOutput, ProtocolResponseError
from deepstrike.providers.stop_reason import canonicalize_stop_reason
from deepstrike.providers.stream import TextDelta, ToolCallEvent, UsageEvent
from deepstrike.providers.usage import ProviderUsage
from deepstrike.types.content import CanonicalAdapterInput, normalize_tool_result, project_tool_output_to_text


def _number(raw: dict, field: str) -> int | None:
    value = raw.get(field)
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)) or value < 0:
        raise ProtocolResponseError("ollama-chat", f"{field} must be a non-negative number")
    return int(value)


class OllamaNdjsonDecoder:
    def __init__(self) -> None:
        self._buffer = ""

    def push(self, text: str) -> list[dict]:
        self._buffer += text
        lines = self._buffer.split("\n")
        self._buffer = lines.pop()
        return self._parse(lines)

    def finish(self, text: str = "") -> list[dict]:
        self._buffer += text
        tail, self._buffer = self._buffer, ""
        return self._parse([tail] if tail else [])

    @staticmethod
    def _parse(lines: list[str]) -> list[dict]:
        chunks = []
        for line in lines:
            if not line.strip():
                continue
            try:
                chunk = json.loads(line)
            except json.JSONDecodeError:
                continue
            if isinstance(chunk, dict):
                chunks.append(chunk)
        return chunks


@dataclass
class OllamaStreamState:
    pending_tool_calls: dict[str, dict] = field(default_factory=dict)
    final_chunk: dict | None = None


class OllamaAdapter:
    protocol = "ollama-chat"

    def __init__(self, model: str = "llama3") -> None:
        self._model = model

    def create_ndjson_decoder(self) -> OllamaNdjsonDecoder:
        return OllamaNdjsonDecoder()

    def build_request(self, input: CanonicalAdapterInput) -> dict:
        context = input.context
        messages: list[dict] = []
        if context.system_text:
            messages.append({"role": "system", "content": context.system_text})
        turns = [*context.turns, *([context.state_turn] if context.state_turn is not None else [])]
        for message in turns:
            entry: dict = {"role": message.role, "content": message.content}
            parts = getattr(message, "content_parts", None) or []
            if any(part.type == "audio" for part in parts):
                raise UnsupportedModalityError("audio", "ollama")
            images = [part.data for part in parts if part.type == "image" and part.data]
            tool_results = [part for part in parts if part.type == "tool_result"]
            if tool_results:
                part = tool_results[0]
                entry["content"] = project_tool_output_to_text(normalize_tool_result(
                    part.call_id, part.output, part.is_error, getattr(part, "content_parts", None),
                ).blocks)
            if images:
                entry["images"] = images
            messages.append(entry)
        request = {
            **{key: value for key, value in input.extensions.items() if key not in {"model", "messages", "tools", "stream"}},
            "model": input.resolved.model_id if input.resolved is not None else self._model_from_context(input) or self._model,
            "messages": messages,
        }
        if input.tools:
            request["tools"] = [{"type": "function", "function": {
                "name": tool.name,
                "description": tool.description,
                "parameters": json.loads(tool.parameters),
            }} for tool in input.tools]
        return request

    @staticmethod
    def _model_from_context(input: CanonicalAdapterInput) -> str:
        return str(input.extensions.get("model") or "")

    def decode_complete(self, raw: dict, input: CanonicalAdapterInput) -> Message:
        message = raw.get("message") or {}
        tool_calls = []
        for call in message.get("tool_calls") or []:
            function = call.get("function") or {}
            normalized = normalize_tool_call(call.get("id", ""), function.get("name", ""), function.get("arguments", {}))
            if normalized:
                tool_calls.append(normalized)
        return Message(role="assistant", content=message.get("content") or "", token_count=0, tool_calls=tool_calls or None)

    def create_stream_state(self, input: CanonicalAdapterInput) -> OllamaStreamState:
        return OllamaStreamState()

    def push_stream_chunk(self, chunk: dict, state: OllamaStreamState) -> AdapterOutput:
        events = []
        message = chunk.get("message") or {}
        if message.get("content"):
            events.append(TextDelta(delta=message["content"]))
        for call in message.get("tool_calls") or []:
            function = call.get("function") or {}
            normalized = normalize_tool_call("", function.get("name", ""), function.get("arguments", {}))
            if normalized is None:
                continue
            key = f"{normalized.name}:{normalized.arguments}"
            if key not in state.pending_tool_calls:
                state.pending_tool_calls[key] = {
                    "id": f"call_{len(state.pending_tool_calls) + 1}",
                    "name": normalized.name,
                    "arguments": json.loads(normalized.arguments),
                }
        if chunk.get("done"):
            state.final_chunk = chunk
        return AdapterOutput(events=events)

    def finish_stream(self, state: OllamaStreamState, final: dict | None = None) -> AdapterOutput:
        terminal = final or state.final_chunk
        events = [ToolCallEvent(**call) for call in state.pending_tool_calls.values()]
        usage = self.normalize_usage(terminal)
        if usage is not None:
            raw_stop_reason = terminal.get("done_reason") if terminal else None
            events.append(UsageEvent(
                total_tokens=usage.input_tokens + usage.output_tokens,
                input_tokens=usage.input_tokens,
                output_tokens=usage.output_tokens,
                stop_reason=canonicalize_stop_reason(raw_stop_reason),
                raw_stop_reason=raw_stop_reason,
                provider_usage=usage,
            ))
        return AdapterOutput(events=events)

    def normalize_usage(self, raw: Any) -> ProviderUsage | None:
        if raw is None:
            return None
        if not isinstance(raw, dict):
            raise ProtocolResponseError("ollama-chat", "usage source must be an object")
        input_tokens = _number(raw, "prompt_eval_count")
        output_tokens = _number(raw, "eval_count")
        if input_tokens is None and output_tokens is None:
            return None
        return ProviderUsage(input_tokens=input_tokens or 0, output_tokens=output_tokens or 0)
